pkg/build/api/types.go

```diff
@@ -352,6 +356,25 @@ type SourceBuildStrategy struct {
 	ForcePull bool
 }
 
+// BuildHookSpec holds build hooks with actions grouped by build status.
+type BuildHookSpec struct {
+	// The OnSuccess group is a list of BuildHooks that will be executed
+	// when the build changes to BuildPhaseComplete.
+	OnSuccess []BuildHook
+
+	// The OnFailure group is a list of BuildHooks that will be executed
+	// when the build changes to BuildPhaseFailed.
+	OnFailure []BuildHook
+}
+
+// A BuildHook defines actions to be executed when a build changes to a certain
+// phase.
+type BuildHook struct {
+	// StartBuilds holds a list of references to BuildConfigs that will be
+	// started when the hook is executed.
+	StartBuilds []kapi.ObjectReference
+}
```

Comment:
> ```go
type Action interface {
   // bool indicates if action should be retried, error indicates the error that occured if any
   func Invoke() bool,error
 }
```

@bparees AFAICT using interface types in the API like that will only bring us trouble. He's an example of using what you proposed: http://play.golang.org/p/p8qJ64PyRi

To work that out we'd need a specialized unmarshaler to convert an `interface{}` into a proper implementation.

Your proposal also removes one level of nesting at the expense of every client using the API having to group actions on their own. We'll need to present the actions in `oc describe` and in the Web Console, and IMO it's cleaner to print actions in logical groups.
I don't see any clear use case of assigning the very same action to multiple build phase transitions, something your design would address.